### PR TITLE
feat(chapter_dynamic_programming): Add js and ts codes for chapter 14.1

### PR DESCRIPTION
**Pull request by yuan0221**
If this PR is related to coding or code translation, please fill out the checklist and paste the console outputs to the PR.

- [x] I've tested the code and ensured the outputs are the same as the outputs of reference codes.
- [x] I've checked the codes (formatting, comments, indentation, file header, etc) carefully.
- [x] The code does not rely on a particular environment or IDE and can be executed on a standard system (Win, macOS, Ubuntu).

### CODE REVIEW

File: codes/javascript/chapter_dynamic_programming/climbing_stairs_backtrack.js [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+function climbingStairsBacktrack(n) {
+  const choices = [1, 2]; // 可选择向上爬 1 或 2 阶
+  const state = 0; // 从第 0 阶开始爬
+  const res = new Map();
+  res.set(0, 0); // 使用 res[0] 记录方案数量
+  backtrack(choices, state, n, res);
+  return res.get(0);
+}
+
+function main() {
```
Comment (krahets) [inline codes/javascript/chapter_dynamic_programming/climbing_stairs_backtrack.js:33]: /* Driver Code */

Save as others

File: codes/javascript/chapter_dynamic_programming/climbing_stairs_backtrack.js [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    return res.get(0);
+}
+
+/* Driver Code */
+function main() {
+    const n = 9;
+    const res = climbingStairsBacktrack(n);
+    console.log(`爬 ${n} 阶楼梯共有 ${res} 种方案`);
+}
+main();
```
Comment (justin-tse) [inline codes/javascript/chapter_dynamic_programming/climbing_stairs_backtrack.js:37]:
```suggestion
const n = 9;
const res = climbingStairsBacktrack(n);
console.log(`爬 ${n} 阶楼梯共有 ${res} 种方案`);
```

Comment (justin-tse) [reply codes/javascript/chapter_dynamic_programming/climbing_stairs_backtrack.js:37]: Save as others, in JS and TS, we do not need to define the main function to execute the code.

File: codes/typescript/chapter_dynamic_programming/climbing_stairs_backtrack.ts [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+ * Author: yuan0221 (yl1452491917@gmail.com)
+ */
+
+/* 回溯 */
+function backtrack(
+    choices: number[],
+    state: number,
+    n: number,
+    res: Map<0, any>
+) {
```
Comment (justin-tse) [inline codes/typescript/chapter_dynamic_programming/climbing_stairs_backtrack.ts:13]:
```suggestion
): void {
```

File: codes/typescript/chapter_dynamic_programming/climbing_stairs_backtrack.ts [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        // 剪枝：不允许越过第 n 阶
+        if (state + choice > n) break;
+        // 尝试：做出选择，更新状态
+        backtrack(choices, state + choice, n, res);
+        // 回退
+    }
+}
+
+/* 爬楼梯：回溯 */
+function climbingStairsBacktrack(n: number) {
```
Comment (justin-tse) [inline codes/typescript/chapter_dynamic_programming/climbing_stairs_backtrack.ts:27]:
```suggestion
function climbingStairsBacktrack(n: number): number {
```

File: codes/typescript/chapter_dynamic_programming/climbing_stairs_dfs.ts [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+function dfs(i: number): number {
+    // 已知 dp[1] 和 dp[2] ，返回之
+    if (i == 1 || i == 2) return i;
+    // dp[i] = dp[i-1] + dp[i-2]
+    const count = dfs(i - 1) + dfs(i - 2);
+    return count;
+}
+
+/* 爬楼梯：搜索 */
+function climbingStairsDFS(n: number) {
```
Comment (justin-tse) [inline codes/typescript/chapter_dynamic_programming/climbing_stairs_dfs.ts:17]:
```suggestion
function climbingStairsDFS(n: number): number {
```

File: codes/typescript/chapter_dynamic_programming/climbing_stairs_dfs_mem.ts [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    if (mem[i] != -1) return mem[i];
+    // dp[i] = dp[i-1] + dp[i-2]
+    const count = dfs(i - 1, mem) + dfs(i - 2, mem);
+    // 记录 dp[i]
+    mem[i] = count;
+    return count;
+}
+
+/* 爬楼梯：记忆化搜索 */
+function climbingStairsDFSMem(n: number) {
```
Comment (justin-tse) [inline codes/typescript/chapter_dynamic_programming/climbing_stairs_dfs_mem.ts:21]:
```suggestion
function climbingStairsDFSMem(n: number): number {
```

**Review by justin-tse — COMMENTED** [review]
@yuan0221 Thank you for your clean code! Please address the following code.

**Comment by yuan0221** [comment]
> 

Done, thx

**Review by justin-tse — APPROVED** [review]
Great, thank you!

**Review by krahets — APPROVED** [review]

**Pull request merged**